Shell add dictionary command

jakub-uC opened the pull request:
Added dictionary commands concept. 

Example usage:

```
static int gain_cmd_handler(const struct shell *shell, size_t argc, char **argv,                                         
                            void *data)                                                                                  
{                                                                                                                        
        int gain;                                                                                                        
                                                                                                                         
        /* data is a value corresponding to called command syntax */                                                     
        gain = (int)data;                                                                                                
                                                                                                                         
        adc_gain_set(gain);                                                                                              
                                                                                                                         
        shell_print(shell, "ADC gain set to: %s\n"                                                                       
                           "Value send to ADC driver: %d",                                                               
                           argv[0],                                                                                      
                           gain);                                                                                        
                                                                                                                         
        return 0;                                                                                                        
}                                                                                                                        
                                                                                                                         
SHELL_SUBCMD_DICT_SET_CREATE(sub_gain, gain_cmd_handler,                                                                 
        (gain_1, 1), (gain_2, 2), (gain_1_2, 3), (gain_1_4, 4)                                                           
);                                                                                                                       
SHELL_CMD_REGISTER(gain, &sub_gain, "Set ADC gain", NULL); 
```

![dict_cmd](https://user-images.githubusercontent.com/34125330/95325248-352f9600-08a1-11eb-8b75-900da9a3a591.png)

nashif commented:
> Added dictionary commands concept.

can you add some more details about this concept? How about describing this in the documentation as well?

Commit:
shell: dictionary commands feature

Adding dictionary commands concept. It simplifies creation of multiple
commands that are handled by a one function handler.

This is a special kind of static commands. Dictionary commands can be
used every time you want to use a pair: (string <-> corresponding data)
in a command handler. The string is usually a verbal description
of a given data. The idea is to use the string as a command syntax that
can be prompted by the shell and corresponding data can be used to
process the command.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: shell example update

Updating shell example to present usage of dictionary commands.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nordic-krch commented on include/shell/shell.h:
```diff
@@ -396,6 +412,52 @@ struct shell_static_entry {
+ *	SHELL_CMD_REGISTER(dictionary, &sub_dict_cmds, NULL, NULL);
+ */
+#define SHELL_SUBCMD_DICT_SET_CREATE(_name_, _handler_, ...)		\
+	FOR_EACH_FIXED_ARG(SHELL_CMD_DICT_HANDLER_CREATE, (;),		\
```
it should rather have empty separator `()` since it is separating functions.

nordic-krch commented on include/shell/shell.h:
```diff
@@ -396,6 +412,52 @@ struct shell_static_entry {
+
+/**
+ * @brief Initializes shell dictionary commands.
+ *
```
Please elaborate here when dictionary commands can be used.

jakub-uC replied:
I am working on documentation now. Once I will have it I will update the description as well.

nordic-krch commented on include/shell/shell.h:
```diff
@@ -139,6 +139,22 @@ const struct device *shell_device_lookup(size_t idx,
 				 size_t argc, char **argv);
 
+/**
+ * @brief Shell command handler prototype.
```
it is rather handler used for dictionary.

Commit:
shell: documentation update

Updating documentation with dictionary commands description.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jakub-uC commented:
> > Added dictionary commands concept.
> 
> can you add some more details about this concept? How about describing this in the documentation as well?

I've updated the documentation and I've added some examples.

nordic-krch reviewed (approved):
looks good

jukkar reviewed (approved):
No objections
